crates/astria-sequencer-relayer/src/data_availability.rs

```diff
@@ -337,6 +337,66 @@ impl CelestiaClient {
         Ok(sequencer_namespace_datas)
     }
 
+    /// Returns the rollup data for a given rollup namespace at a given height, if it exists.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// + the verification key could not be constructed from the data stored in `namespace_data`;
+    /// + the RPC to fetch the blobs failed.
+    pub async fn get_rollup_data(
+        &self,
+        height: u64,
+        namespace_data: &SignedNamespaceData<SequencerNamespaceData>,
+        rollup_namespace: Namespace,
+    ) -> eyre::Result<Option<RollupNamespaceData>> {
+        let verification_key = VerificationKey::try_from(&*namespace_data.public_key)
+            .wrap_err("failed constructing verification key from stored bytes")?;
+
+        let req = GetAllRequest {
+            height,
+            namespace_ids: vec![*rollup_namespace],
+        };
+
+        let rsp = match self.client.blob_get_all(req).await {
+            Ok(rsp) => rsp,
+            Err(e) => {
+                if let ErrorKind::Rpc(astria_celestia_jsonrpc_client::JsonRpseeError::Call(inner)) =
+                    e.kind()
+                {
+                    if inner.message().contains("blob: not found") {
+                        info!("could not find blobs under the listed namespaces");
+                        return Ok(None);
+                    }
+                }
+                return Err(e).wrap_err("failed getting namespaced data");
+            }
+        };
+
+        // filter out blobs; we should only be left with either zero or one rollup datas
+        let rollup_datas = filter_and_convert_rollup_data_blobs(
+            rsp.blobs,
+            &namespace_data.data.block_hash,
+            &verification_key,
+        );
+
+        // this should *not* happen; the only case where it would happen is if the sequencer-relayer
+        // posts multiple blobs with the same rollup ID for the same block (could this happen?)
+        if rollup_datas.len() > 1 {
+            warn!(
+                num_rollup_datas = rollup_datas.len(),
+                "found more than one rollup data for the given block hash; returning the first one"
+            );
+        }
+
+        // this case can happen if someone posts a blob to the namespace with invalid data
+        if rollup_datas.is_empty() {
+            return Ok(None);
+        }
+
+        Ok(rollup_datas.into_iter().next().map(|data| data.1.data))
```

Comment:
Re my comment above: assuming something bad like this happens, I would want to understand what namespaced data was matched.